[BugFix] Fix snapshot was deleted by mistake after restore from it

srlch opened the pull request:
## Why I'm doing:
Suppose current cluster has enabled cluster snapshot and has generated twice says snapshot_a and snapshot_b so far. When the snapshot_b has been finished, the final state for two snapshot jobs should be: snapshot_a -> DELETED, snapshot_b -> FINISHED. But the state for this two job in the backuped snapshot in remote is: snapshot_a -> FINISHED, snapshot_b -> INITIALIZING, this because some state transition editlog do not included (by designed, not a bug). So if we use this snapshot to do the restoring, after restore, snapshot_a will be FINISHED and snapshot_b will be forced to set error and delete the snapshot in the remote storage.

## What I'm doing:
Reset the snapshot we use to restore if the relative job state is not finished to prevent the deletion.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -224,11 +224,29 @@ public NavigableMap<Long, ClusterSnapshotJob> getAutomatedSnapshotJobs() {
         clearFinishedAutomatedClusterSnapshotExceptLastFinished();
     }
 
+    public void setLastJobFinishedIfRestoredFromIt(String restoredSnapshotName) {
```
```suggestion
    public void setLastJobFinishedAfterRestored(String restoredSnapshotName) {
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -224,11 +224,29 @@ public NavigableMap<Long, ClusterSnapshotJob> getAutomatedSnapshotJobs() {
+    public void resetAutomatedJobsStateForTheFirstRun(String restoredSnapshotName) {
+        setLastJobFinishedIfRestoredFromIt(restoredSnapshotName);
         resetLastUnFinishedAutomatedSnapshotJob();
         clearFinishedAutomatedClusterSnapshotExceptLastFinished();
```
```suggestion
        clearFinishedAutomatedClusterSnapshotExceptLast();
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -224,11 +224,29 @@ public NavigableMap<Long, ClusterSnapshotJob> getAutomatedSnapshotJobs() {
+            // editlog for the state transtition after ClusterSnapshotJobState.INITIALIZING
+            if (job.getSnapshotName().equals(restoredSnapshotName) && job.isInit()) {
+                job.setState(ClusterSnapshotJobState.FINISHED);
+                job.setDetailInfo("Finished time was reset after cluster restore");
```
```suggestion
                job.setDetailInfo("Finished time was reset after cluster restored");
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotJob.java:
```diff
@@ -118,6 +121,10 @@ public boolean isUnFinishedState() {
                 state == ClusterSnapshotJobState.UPLOADING;
     }
 
+    public boolean isInit() {
```
```suggestion
    public boolean isInitializing() {
```

srlch replied:
done

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -224,12 +225,37 @@ public NavigableMap<Long, ClusterSnapshotJob> getAutomatedSnapshotJobs() {
     }
 
-    public void resetAutomatedJobsStateForTheFirstRun() {
+    public void resetAutomatedJobsStateForTheFirstRun(RestoredSnapshotInfo restoredSnapshotInfo) {
```
```suggestion
    public void resetSnapshotJobsStateAfterRestarted(RestoredSnapshotInfo restoredSnapshotInfo) {
```

srlch replied:
done

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -224,12 +225,37 @@ public NavigableMap<Long, ClusterSnapshotJob> getAutomatedSnapshotJobs() {
+            return;
+        }
+
+        if (!automatedSnapshotJobs.isEmpty()) {
+            ClusterSnapshotJob job = automatedSnapshotJobs.lastEntry().getValue();
```
not atomic

srlch replied:
done

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/RestoredSnapshotInfo.java:
```diff
@@ -0,0 +1,36 @@
+// save the infomation parsed from the cluster snapshot when start in restore mode
+public class RestoredSnapshotInfo {
+    private String restoredSnapshotName;
+    private Pair<Long, Long> restoredJouranlIds;
```
```suggestion
        private long feJournalId;
        private long starMgrJournalId;
```

srlch replied:
done

srlch committed:
## Why I'm doing:
Suppose current cluster has enabled cluster snapshot and has generated twice says snapshot_a and snapshot_b so far. When the snapshot_b has been finished, the final state for two snapshot jobs should be: snapshot_a -> DELETED, snapshot_b -> FINISHED. But the state for this two job in the backuped snapshot in remote is: snapshot_a -> FINISHED, snapshot_b -> INITIALIZING, this because some state transition editlog do not included (by designed, not a bug). So if we use this snapshot to do the restoring, after restore, snapshot_a will be FINISHED and snapshot_b will be forced to set error and delete the snapshot in the remote storage.

## What I'm doing:
Reset the snapshot we use to restore if the relative job state is not finished to prevent the deletion.

Signed-off-by: srlch <linzichao@starrocks.com>

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/RestoredSnapshotInfo.java:
```diff
@@ -0,0 +1,40 @@
+
+// save the infomation parsed from the cluster snapshot when start in restore mode
+public class RestoredSnapshotInfo {
+    private String restoredSnapshotName;
```
Since there is `restored` in class name, no need to add it in member name again.

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/RestoredSnapshotInfo.java:
```diff
@@ -0,0 +1,40 @@
+        return feJournalId;
+    }
+
+    public long getRestoredStarMgrJournalIds() {
```
```suggestion
    public long getStarMgrJournalId() {
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -32,6 +32,8 @@ public class ClusterSnapshotCheckpointScheduler extends FrontendDaemon {
 
     private final CheckpointController feController;
     private final CheckpointController starMgrController;
+    // cluster snapshot infomation used for start
```
```suggestion
    // cluster snapshot information used for start
```

srlch replied:
updated

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/RestoreClusterSnapshotMgr.java:
```diff
@@ -131,6 +143,36 @@ private void downloadSnapshot() throws StarRocksException {
+
+            LOG.info("Download cluster snapshot successfully with FE image version: {}, StarMgr image version: {}",
+                     feImageJournalId, starMgrImageJournalId);
+        } catch (IOException e) {
+            LOG.error("Failed to get image version info for cluster snapshot", e);
+            return;
+        }
```
throw a StarRocksException to let fe start failed.

srlch replied:
updated

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/RestoredSnapshotInfo.java:
```diff
@@ -0,0 +1,40 @@
+        this.starMgrJournalId = starMgrJournalId;
+    }
+
+    public String getRestoredSnapshotName() {
```
updated

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/RestoredSnapshotInfo.java:
```diff
@@ -0,0 +1,40 @@
+        return restoredSnapshotName;
+    }
+
+    public long getRestoredFEJournalIds() {
```
updated
